<div style="width: 100%; clear: both;">
    <div style="float: left; width: 50%;">
       <img src="http://www.uoc.edu/portal/_resources/common/imatges/marca_UOC/UOC_Masterbrand.jpg", align="left">
    </div>
</div>

<div style="float: right; width: 50%;">
    <p style="margin: 0; padding-top: 22px; text-align:right;">22.403 · Programació per a la ciència de dades</p>
    <p style="margin: 0; text-align:right;">Grau en Ciència de Dades Aplicada</p>
    <p style="margin: 0; text-align:right; padding-button: 100px;">Estudis d'Informàtica, Multimèdia i Telecomunicació</p>
</div>

</div>
<div style="width: 100%; clear: both;">
<div style="width:100%;">&nbsp;</div>


Programació per a la ciència de dades - PAC2
============================

En aquest Notebook trobareu un conjunt d'exercicis que suposen la segona activitat d'avaluació continuada (PAC) de l'assignatura.

Per a cada exercici, tingueu en compte que:
* **És necessari incloure comentaris** del vostre codi, que expliquin com heu implementat la solució del problema plantejat.
* **És imprescindible** citar les referències que heu consultat per a dur a terme l'activitat.
* Es valorarà que el codi proporcionat solucioni el problema proposat i també la qualitat del codi (comentaris, llegibilitat, claredat, ús de les estructures de dades adequades, bona nomenclatura de les variables i funcions, seguiment del PEP8, etc.).

Veureu que totes les activitats tenen una etiqueta, que indica els recursos necessaris per tal de dur-la a terme. Hi ha tres possibles etiquetes:

* <span style="font-family: Courier New; background-color: #82b74b; color: #000000; padding: 3px; ">NM</span> **Només materials**: les eines necessàries per a realitzar l'activitat es poden trobar als materials de l'assignatura (considerarem també els materials de l'assignatura Fonaments de Programació, així com les lectures obligatòries de material extern que s'indiquen als notebooks).

* <span style="font-family: Courier New; background-color: #ffcc5c; color: #000000; padding: 3px; ">EG</span> **Consulta externa guiada**: l'activitat pot requerir fer ús d'eines que no es troben als materials de l'assignatura, però l'enunciat conté indicacions d'on o com trobar la informació addicional necessària per resoldre l'activitat.

* <span style="font-family: Courier New; background-color: #f2ae72; color: #000000; padding: 3px; ">EI</span> **Consulta externa independent**: l'activitat pot requerir fer ús d'eines que no es troben als materials de l'assignatura, i l'enunciat pot no incloure la descripció d'on o com trobar aquesta informació addicional. Caldrà que l'estudiant busqui aquesta informació fent servir els recursos que s'han explicat a l'assignatura.

És important notar que aquestes etiquetes no indiquen el nivell de dificultat de l'exercici, sinó únicament la necessitat de consulta de documentació externa per a la seva resolució. A més, recordeu que les **etiquetes són informatives**, però **podeu consultar referències externes, i usar-les, sempre que vulgueu (encara que no s'indiqui explícitament) o pot ser que pugueu fer una activitat sense consultar cap mena de documentació**. Per exemple, per resoldre una activitat que només requereixi els materials de l'assignatura, podeu consultar referències externes si voleu, ja sigui tant per ajudar-vos en la seva resolució com per ampliar el coneixement!


**Important**: A més de la puntuació de cada exercici, que valora la qualitat de la solució que proposeu i els resultats obtinguts, en aquesta PAC es valoren 3 aspectes addicionals, transversals a totes les activitats:

* **Documentació** (**0.25*f punts**): Totes les funcions dels exercicis d’aquesta PAC hauran de ser correctament documentades utilitzant docstrings (en el format que preferiu).

* **Modularitat** (**0.5*f punts**): Es valorarà la modularitat del codi (la correcta estructuració del codi en funcions).

* **Estil** (**0.25*f punts**): El codi ha de seguir la guia d'estil de Python [PEP8](https://www.python.org/dev/peps/pep-0008/), exceptuant els casos on fer-ho compliqui la llegibilitat del codi.

On $f$ es la fracció de la PAC realitzada. És a dir, la nota d'aquests aspectes transversals serà proporcional a la part de la PAC realitzada.

Així, els aspectes transversals tindran una puntuació màxima d'1 punt si es lliuren tots els exercicis de la PAC. En cas contrari, la nota d'aspectes transversals es multiplicarà per la fracció de la PAC lliurada. Per exemple, si els aspectes transversals estan puntuats amb un 1 i només hi ha la meitat de la PAC lliurada, la nota dels aspectes transversals serà $1*0.5=0.5$.

La nota dels aspectes transversals es considera inclosa en les notes mencionades per als exercicis.


---

## Abans de començar

Si us plau, llegiu atentament els següents paràgrafs relatius a l'originalitat en les activitats de l'assignatura abans de començar a treballar en la PAC. Si teniu qualsevol dubte, dirigiu-vos al professor col·laborador de l'aula abans de seguir amb l'activitat.

<div class="alert alert-block alert-success">La manca d’originalitat es produeix quan en una activitat hi apareix contingut que no ha estat elaborat de forma individual per l’estudiant i no es referència adequadament el seu origen. O bé quan, encara que el contingut extern estigui referenciat, aquest és tan extens que no és possible considerar a l’estudiant l’autor o autora de l’activitat.
<br /><br />
Així, alguns exemples de comportaments inadequats són pel que fa a la manca d'originalitat són:
<ol>
    <li>Crear la solució d'un exercici de la PAC en col·laboració entre diversos estudiants.</li>
    <li>Incloure un exercici de la PAC que utilitzi codi trobat a internet sense citar-ne la font.</li>
    <li>Compartir la vostra solució de la PAC amb altres estudiants de l'assignatura.</li>
    </ol>
</div>

## Instruccions de lliurament

Per tal de lliurar l'activitat és necessari que feu els següents passos:

1. Comproveu que el notebook s'executa correctament a google colab. És important que abans de lliurar la vostra PAC comproveu que la versió final del codi que lliureu s'executa correctament en la seva totalitat. Per fer-ho, us recomanem fer una execució completa des de zero del notebook: feu click a *Runtime > Disconnect and Delete Runtime & Run All* i comproveu que el notebook s'executa correctament.
2. Activeu les alertes d'estil i comproveu que el codi segueix les convencions d'estil de Python. Si és necessari, modifiqueu el codi per adequar-lo.
3. Confirmeu que sou els autors únics de la PAC i que aquesta inclou totes les cites a recursos externs que s'hagin utilitzat per a elaborar-la. Per tal de confirmar que sou els autors únics de l'activitat, afegiu el vostre nom complet al paràgraf de la cel·la següent.
4. Per a aquesta PAC només caldrà que lliureu el fitxer amb el notebook (amb extensió .ipynb). Nosaltres hi afegirem la carpeta "data".
5. Lliureu la PAC a través de l'aula: "Activitats,Lliurament de l'activitat PAC1".
6. La càrrega del dataset s'ha de fer usant **rutes relatives**, heu de carregar el fitxer tenint en compte que treballeu (el vostre notebook es troba) en una carpeta que conté la carpeta data i l'arxiu de dades corresponent.

<div class="alert alert-block alert-warning">Jo, Marc Bracons Cucó, confirmo que he elaborat de manera individual totes les activitats resoltes en aquesta PAC i que he inclòs les cites a totes les fonts externes que he utilitzat per a resoldre les activitats. </div>

---

**NOTES IMPORTANTS:**

1) Durant la PAC comptareu amb cel·les de tests per a comprovar que els resultats del vostre codi són correctes. La intenció d'aquestes cel·les és ajudar-vos a la resolució de la PAC, però no es faran servir directament en la correcció. En alguns casos poc freqüents, pot ser que algun codi correcte no passi el test de la cel·la o que algun d'incorrecte sí que el passi. Feu servir el sentit comú i adreceu-vos al vostre consultor si teniu dubtes.

2) En diversos punts de la PAC se us demana de mostrar per pantalla alguna informació o estructura de dades. Recordeu de formatar de forma correcta la sortida en tots els casos.

3) Cal resoldre tots els exercicis de la PAC evitant recorreguts innecessaris de les dades i fent servir, tant com sigui possible, programació funcional i ús avançat de funcions. **Per obtenir la màxima puntuació en els exercicis cal que el codi que feu segueixi estrictament aquests principis a més dels aspectes transversals ja mencionats.**

---

In [1]:
# instalem guia estil
!pip install pycodestyle pycodestyle_magic
!pip install flake8-nb
%load_ext pycodestyle_magic

In [2]:
#activem guia estil
%load_ext pycodestyle_magic
%pycodestyle_on

The pycodestyle_magic extension is already loaded. To reload it, use:
  %reload_ext pycodestyle_magic


In [3]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [4]:
# Copy the original from your drive into the colab machine, to make file handling faster.
%cp /content/drive/MyDrive/Colab_Notebooks/activity_2/original.zip ./
#%cp /content/drive/MyDrive/Colab Notebooks/activity_2/original.zip ./

INFO:pycodestyle:1:80: E501 line too long (89 > 79 characters)
INFO:pycodestyle:2:6: E225 missing whitespace around operator
INFO:pycodestyle:3:1: E265 block comment should start with '# '


# Exercici 1

Responeu si són Certes o Falses les següents preguntes i raoneu breument la vostra resposta:

<span style="font-family: Courier New; background-color: #82b74b; color: #000000; padding: 3px; ">NM</span> **(1 punt)**


*(a) Per tal de fer servir una variable global dins d'una funció cal fer servir la sentència "nonlocal".*

**Resposta:** Fals. "El nonlocal té una funcionalitat similar al global que havíem vist en la funció fun_4, però serveix per referir-se a una variable definida en un àmbit superior no global." (2-CAT-Ús avançat de funcions)

In [5]:
x = 5


def exemple_global():
    global x
    x = 2  # Modifiquem la variable de forma global


exemple_global()
print(x)  # imprimirà 2


def f_externa():
    x = 70

    def f_interna():
        nonlocal x
        x = 50  # Modifiquem la variable dintre de f_interna i f_externa

    f_interna()
    print(x)  # imprimirà 50


f_externa()
print(x)  # imprimirà 2, ja que no hem fet una modificació global
#          al fer "nonlocal x"

2
50
2


*(b) Considerem una variable "float_var" que conte valors de tipus real.

Una funció que tingui per darrera linia

print(float_var)

retornara un valor de tipus string.
.*

**Resposta:** Fals. Si *float_var* és una variable de tipus float, llavors *print(float_var)* mostrarà aquest valor en forma de text a la pantalla, pero no canvia el tipus de valor en si. Per convertir un float a un string en Python hem de fer servir la funció *str()*.



In [6]:
float_var = 3.14


def imprimir_valor():
    print(float_var)  # Això imprimeix el valor float_var a la pantalla
    print("El tipus de float_var és:", type(float_var))


imprimir_valor()


def imprimir_string():
    # Això imprimeix el valor float_var convertit a string
    # a la pantalla
    print(str(float_var))
    print("El tipus de str(float_var) és:",
          type(str(float_var)))


imprimir_string()

3.14
El tipus de float_var és: <class 'float'>
3.14
El tipus de str(float_var) és: <class 'str'>


*(c) Les funcions de Python es poden passar com a paràmetres i ser retornades per altres funcions*.

**Resposta:** Cert. A Python les funciones poden passar com a paràmetres a altres funcions, ser retornades per altres funcions i fins i tot ser assignades a variables.

In [7]:
def saluda(nom):
    return f'Hola, {nom}'


def crida_f(func, argument):
    return func(argument)


print(crida_f(saluda, 'Yago'))

Hola, Yago


(d) Si obrim un fitxer fent servir el mode `w+` mantindrem el contingut del fitxer i es generarà un error si el fitxer no existeix.

**Resposta:** Fals. El mode *w* permet escriure en un fixter, però en sobreescriu el contingut. En mode d'actualització, *+*, permet llegir i escriure un fitxer. *w+* sobreescriurem el contingut del fitxer i crearem el fitxer si aquest no existeix. (3-CAT_Fitxers_i_interacció_amb_el_sistema)

## Exercici 2

Donades notícies recents insinuant que alguns governs oculten secrets sobre possibles artefactes extraterrestres, el cap de l'empresa de data science on treballeu ha decidit de crear un projecte de "fact finding" sobre el tema. Com a primer resultat, altres membres de l'equip han aconseguit de recopilar informació sobre avistaments d'OVNIs a Estats Units entre el 1969 i el 2021. Dissortadament, aquests companys no havien cursat aquesta assignatura i l'organització de dades que han fet no és prou correcta.

En aquest exercici farem servir els coneixements sobre gestió de fitxers i programació funcional que estem aprenent per a llegir l'estructura de carpetes existent i transformar-la en un dataframe de pandas. En un exercici posterior considerarem una altra estructura de fitxers que el cap considera millor i hi guardarem les dades. Durant tot el procés, estarem particularment atents a l'espai ocupat per les dades.

Hem rebut les dades en forma d'un sol fitxer comprimit anomenat **original.zip**.

## 2.1

Donat que ens interessa la mida que els diversos fitxers i carpetes ocupen a memòria, cal que **creeu una funció per a mesurar la mida de fitxers o carpetes**. Cal que aquesta funció executi correctament tant si li passem la ruta d'un fitxer com d'un directori. **Descomprimiu també el fitxer original.zip per codi**. **Cal que la funció retorni la mida en Megabytes**.

Com a prova, mesureu la mida del fitxer **original.zip** i de l'estructura de carpetes que obteniu en descomprimir-lo.


## 2.2

Com veureu, l'estructura té una sèrie de carpetes amb fitxers a dins.

En concret:


```
<original>
    <state_code1>
        <city_name1>
            shape1.txt
            shape2.txt
            ...
            shapeN.txt
        <city_name2>
        ...
    <state_code2>
    ...
        
```

on:

- El directori arrel (anomenat "original") conte una subcarpeta per a cada estat d'Estats Units on s'han detectat avistaments.
- **state_code** és un codi de dues lletres per a cadascun dels estats d'Estats Units (per exemple CA representa California o AL Alabama).
- Dins de la carpeta de cada estat hi ha una carpeta **city_name** amb el nom de cada ciutat on s'han detectat avistaments. Per exemple "Chicago".
- A dins de cada carpeta de ciutat, hi haurà un fitxer de text per a cada tipus d'objecte avistat. Per exemple, "circle.txt", "cylinder.txt" o "cigar.txt".
- A cada línia de cada fitxer hi trobareu informació (separada per punt i coma ";" referent als següents conceptes):

summary;date_time;duration;text;city_latitude;city_longitude

- Els noms d'aquests conceptes no apareixen als fitxers, queden implícits per l'ordre a cada línia. Cada fitxer tindrà com a mínim una línia, però pot ser que en tingui més.

**Creeu una funció que rebi com a entrada un string amb la ruta a una carpeta (contenint una estructura com l'esmentada) i retorni un diccionari** amb el següent format:

- Les claus del diccionari seran enters representant l'ordre en què s'ha trobat cadascuna de les línies de cadascun dels arxius "shapei.txt" mencionats anteriorment. Depenent de la implementació pot ser que llegiu els fitxers en ordres diferents.

- El valor associat a cada clau serà, en si mateix, un diccionari amb una clau per a cadascun dels conceptes següents: summary,city,state,date_time,shape,duration,text,city_latitude,city_longitude i el valor associat a cada clau segons consta a l'arxiu que s'està llegint.

**Exemple:**

En el cas que haguéssim llegit com a primer arxiu "disk.txt" dins de la carpeta "IL" i subcarpeta "Chicago", que conté com a primera línia:

"Silver metallic disc observed at Roseland little league baseball field.;1982-06-19 15:00:00;45 seconds;Silver metallic disc observed at Roseland little league baseball field. On a clear...to tell my  story.;41.8841229180932;-87.6674748940178"

Llavors el diccionari donaria a aquesta entrada la clau "0" (primera posició en la seqüència) i tindria, per tant, com a parella (clau, valor):

(0, {'summary': 'Silver metallic disc observed at Roseland little league baseball field.',
  'city': 'Chicago',
  'state': 'IL',
  'date_time': '1982-06-19 15:00:00',
  'shape': 'disk',
  'duration': '45 seconds',
  'text': "Silver metallic disc observed at Roseland little league baseball field. On a clear...to tell my  story.",
  'city_latitude': '41.8841229180932',
  'city_longitude': '-87.6674748940178'})


Noteu que part de la informació (shape, city, state) l'hem obtingut de l'estructura de directoris i la resta del fitxer "disk.txt". Noteu també que hem omès part del camp "text" per a millorar la llegibilitat de la PAC.

**important** Per a resoldre aquest apartat (2.2) **No es pot fer servir la llibreria Pandas.** El seu us resultarà en la pèrdua de tota la puntuació de l'apartat.

## 2.3

- Convertiu el diccionari obtingut en un dataframe de pandas.  
- Guardeu aquest dataframe en un fitxer en format ".csv" anomenat **fullDF.csv**.  
- Mostreu per pantalla les 10 primeres entrades d'aquest dataframe.  
- Mesureu i mostreu per pantalla la mida del fitxer **fullDF.csv**.
- Compareu les diferències en mida entre "original.zip", la carpeta descomprimida i "fullDF.csv" fent servir la funció de l'exercici 2.1. Quin ocupa menys espai. Donat que tots tres contenen essencialment la mateixa informació, quin n'es el motiu? Repetiu l'exercici però ara feu-ho mesurant l'espai en disc que us dona el sistema operatiu. Hi ha diferències? Per quin motiu? Responeu aquestes preguntes en un petit paràgraf de text (<10 línies).

**Nota:** Per saber les mides de fitxers "al sistema operatiu" el més fàcil és que copieu els fitxers i carpetes al vostre drive i els descarregueu per mesurar-los amb les eines del vostre sistema. Aquesta part es considerarà "extra" i es pot obtenir la màxima puntuació a la PAC sense resoldre-la.

<span style="font-family: Courier New; background-color: #82b74b; color: #000000; padding: 3px; ">NM</span> **(4 punts)**

**Solució** En l'apartat 2.1 primer fem una funció per a calcular la mida d'un fitxer o els fitxers d'un directori en Megabytes. A l'exercici no es demana però he pensat que si estem parlant de l'ordre de Megabytes, no té sentit que el resultat tingui molts decimals, així que he fet servir round(x,1) per a canviar-ho.

Pel que fa a l'estructura del directori, he fet una funció per a veure els nivells de les carpetes fins a arribar als fitxers. Com hi ha molts estats, ciutats i fitxers, he modificar la funció per a que només se'n mostrin uns exemples. Aquesta part no es demana a l'exercici però ajuda a entendre amb el que estem treballant i com encarar altres apartats. En aquesta funció també aprofito per calcular el nombre de fixters.

In [8]:
import os
import zipfile


def calcular_mida(ruta: str) -> float:
    """
    Calcula la mida (en megabytes) d'un fitxer o d'un directori

    Args:
    ruta (str): La ruta del fitxer o directori

    Returns:
    float: La mida del fitxer o directori en megabytes
    """
    total_size = 0

    # Si és un fitxer, calcula la dimensió
    if os.path.isfile(ruta):
        total_size = os.path.getsize(ruta)
    # En cas contrari, recorra tots els directoris i sub-directoris
    # root és el camí al directori que s'està examinant
    # dirs és la llista de sub-directoris dintre de root
    # files és la llista de fitxers dintre de root
    else:
        for root, dirs, files in os.walk(ruta):
            for f in files:
                fp = os.path.join(root, f)
                if os.path.isfile(fp):
                    total_size += os.path.getsize(fp)

    return round(total_size / (1024 * 1024), 1)  # Només 1 decimal


def descomprimir_zip(ruta_zip: str, ruta_desti: str) -> None:
    """
    Descomprimeix un fitxer .zip a la ruta de destinació especificada

    Args:
    ruta_zip (str): La ruta del fitxer .zip a descomprimir
    ruta_desti (str): La ruta de destinació on es descomprimiran els arxius

    Returns:
    None
    """
    # Creem una instància de la classe ZipFile, ruta_zip és la ruta del fixter
    # 'r' indica que obrim en mode de lectura (read)
    with zipfile.ZipFile(ruta_zip, 'r') as zip_fitxer:
        # fem servir extractall per a extreure tots els fitxers i carpetes
        # dintre de zip_fitxer a 'ruta_desti' (si no existeix, es crearà)
        zip_fitxer.extractall(ruta_desti)


size_original = calcular_mida('original.zip')
print(f'La mida abans de descomprimir és {size_original} Megabytes')

descomprimir_zip('original.zip', '')
size_decompressed = calcular_mida('original')
print(f'La mida després de descomprimir és {size_decompressed} Megabbytes')

La mida abans de descomprimir és 62.2 Megabytes
La mida després de descomprimir és 111.4 Megabbytes


In [9]:
from collections import defaultdict


def mostrar_estructura(ruta: str, max_per_nivell=2) -> None:
    """
    Mostra l'estructura de directoris d'un directori, limitant el nombre de
    directoris i fitxers mostrats per nivell. També calcula el nombre total
    de fixters en el directori

    Args:
    ruta (str): La ruta del directori
    max_per_nivell (int): Màxim nombre d'elements a mostrar per nivell
                          (2 per defecte)
    Returns:
    None: Imprimeix l'estructura de directoris
    """
    print(f'Estructura de: {ruta}')
    # inicialitzem 'estructura' i 'total_fitxers'. Per a 'estructura'
    # fem servir 'defaultdic' per donar un valor per defecte per a una
    # clau que encara no existeix
    estructura = defaultdict(list)
    total_fitxers = 0

    # root és el camí al directori que s'està examinant
    # dirs és la llista de sub-directoris dintre de root
    # files és la llista de fitxers dintre de root
    for root, dirs, files in os.walk(ruta):
        # relpath és per obtindre el path relatiu de root a ruta
        path = os.path.relpath(root, ruta)
        # os.path.relpath retorn:
        # - '.' si root i ruta són iguals
        # si això passa, path és una cadena buida
        if path == '.':
            path = ''
        # el diccionari 'estructura' s'actualitza:
        # - clau és el 'path' i el valor són els fixters
        #   (màx max_per_nivell) valors
        estructura[path] = dirs[:max_per_nivell] + files[:max_per_nivell]
        # guardem a 'total_fitxers' el nombre de 'files'
        total_fitxers += len(files)

    def imprimir_estructura(actual: str, idx=0):
        """
        Funció interna de llistar_estructura_limitada() que s'utilitza
        per imprimir l'estructura de directoris i fitxers

        Args:
        actual (str): cadena que representa el camí actual
        idx (int): enter per controlar el nivell del directori
        """
        # recorrem tots els elements dintre de 'actual'
        for element in estructura[actual]:
            # impirmim 'element' i fem servir 'idx' per donar un format
            # fàcil per veure el nivell dels directoris
            print('  ' * idx + element)
            # contruïm nou camí afegint 'element' a 'actual'
            # si 'actual' és una cadena buida llavors
            # es pren 'element' com 'nou_actual'
            nou_actual = os.path.join(actual, element) if actual else element
            if nou_actual in estructura:
                imprimir_estructura(nou_actual, idx + 1)

    imprimir_estructura('')
    print(f'Nombre total de fitxers: {total_fitxers}')


mostrar_estructura('original')

Estructura de: original
CO
  Victor
    unknown.txt
    cylinder.txt
  Pueblo West
    triangle.txt
    unknown.txt
NT
  Hay River
    circle.txt
  Fort Mcpherson
    light.txt
Nombre total de fitxers: 49543


In [10]:
# Test Cell, Public
assert 61.8 <= size_original <= 62.4, "Wrong Size for original file."
assert 111 <= size_decompressed <= 112, "Wrong Size for decompressed folder."

**Solució** En l'apartat 2.2 fem una funció que llegeix tots els fitxers d'una carpeta amb l'estructura esmentada i en fa un diccionari. A l'enunciat no es demana però faig una comprovació sobre els fitxers per assegurar que son .txt. Una altre comprovació interessant podria ser mirar si tots els fitxers .txt tenen 9 camps o si n'hi ha algun que no els té (podria ser un fitxer corrupte).

In [11]:
from posixpath import split


def llegir_fitxers_fer_diccionari(ruta):
    """
    Llegeix tots els fitxers acabats en '.txt' en una estructura de
    directoris i retorna un diccionari amb les dades.

    Args:
    ruta (str): La ruta del directori arrel on començar la recerca

    Returns:
    dict: Un diccionari amb les dades extretes
    """
    dades = {}
    id_linia = 0

    # recorrem tots els directoris
    for root, dirs, files in os.walk(ruta):
        # fem un split per partir la ruta i obtenir city i state
        # root conté la ruta completa al fitxer, al fer split respecte
        # 'os.sep' estem separant per directoris (en windows seria '\\'
        # i en Ubuntu '/')
        parts_path = root.split(os.sep)
        # Comprovem si la ruta té la profunditat esperada ja que també passarem
        # pel cas 'original', 'original/estat'
        # abans d'arribar a 'original/estat/ciutat'
        if len(parts_path) >= 3:
            # El codi de l'estat és el tercer des de l'últim
            state = parts_path[-2]
            # El nom de la ciutat és el segon des de l'últim
            city = parts_path[-1]

            # recorrem tots els fitxers
            for file in files:
                # Comprovem si el fitxer és un .txt abans de processar-lo
                if file.endswith('.txt'):
                    # obtenim 'shape' fent un split del nom del fitxer
                    shape = file.split('.txt')[0]
                    # creem la ruta completa del fitxer
                    file_path = os.path.join(root, file)
                    # obrim el fitxer en mode lectura
                    with open(file_path, 'r') as file:
                        # llegim les linies del fitxer
                        for line in file:
                            # fem servir "split(';')" per
                            # separar els diferents camps
                            camps = line.strip().split(';')
                            # Comprovem que tenim 6 camps esperats
                            if len(camps) == 6:
                                # fem servir 'id_linia' per
                                # anar omplint amb els 'camps' del txt
                                dades[id_linia] = {
                                    'summary': camps[0],
                                    'city': city,
                                    'state': state,
                                    'date_time': camps[1],
                                    'shape': shape,
                                    'duration': camps[2],
                                    'text': camps[3],
                                    'city_latitude': camps[4],
                                    'city_longitude': camps[5]
                                }
                                id_linia += 1

    return dades


myDict = llegir_fitxers_fer_diccionari('original')

In [12]:
# Test Cell, Public
dict_size = len(myDict)

assert dict_size == 101207, " Incorrect number of entries."
assert len(myDict[0]) == len(myDict[dict_size-1]) == 9, \
 " Incorrect number of entries."

**Solució** En l'apartat 2.3 hem de transformar el diccionari a un dataframe de pandas. Posteriorment el volem guardar en un fitxer en format csv anomenat fullDF.csv. Per fer-ho he fet una funció que realitza les dues tasques. Els inputs de la funció són el diccionari i el nom del fixter csv, el output és el dataframe per si hi volem treballar.

In [13]:
import pandas as pd
from typing import Dict


def diccionari_to_csv(diccionari: Dict, nom_csv: str) -> pd.DataFrame:
    """
    Converteix un diccionari en un DataFrame de pandas i
    després el guarda com a fitxer CSV.

    Args:
        diccionari (Dict): El diccionari per convertir.
        nom_csv (str): El nom del fitxer CSV de sortida.

    Returns:
        pd.DataFrame: El DataFrame creat a partir del diccionari.
    """
    # Convertim el diccionari en una llista
    data = list(diccionari.values())

    # Creem un dataframe a partir de la llista
    df = pd.DataFrame(data)

    # Guardem el DataFrame com a fitxer CSV
    # afegim index=False per evitar guardar l'index com una columna adicional
    df.to_csv(f'{nom_csv}.csv', index=False)

    # Retornem el DataFrame
    return df


# guardem el dataframe amb el nom 'df_pac2'
df_pac2 = diccionari_to_csv(myDict, 'fullDF')
# mirem les 10 primeres files
print(df_pac2.head(10))

                                             summary         city state  \
0  Two strobe like lights flying in unison in sou...       Victor    CO   
1                  colored lights behind Pike's Peak       Victor    CO   
2  It was about 11:30 at night, when my brother c...  Pueblo West    CO   
3  two bright white lights with a translucent vei...  Pueblo West    CO   
4  Three flashing lights in a triangle shape over...  Pueblo West    CO   
5  Lights Over Lake Pueblo travel past the North ...  Pueblo West    CO   
6      A red laser-like light came out of a cloud...  Pueblo West    CO   
7  Intense red light, spinning rapidly, droning s...  Pueblo West    CO   
8  Red lights that were hovering in the north.  2...  Pueblo West    CO   
9  I have been watching documentaries and studyin...  Pueblo West    CO   

             date_time     shape         duration  \
0  2012-01-26 19:00:00   unknown       15 minutes   
1  2000-09-24 20:15:00  cylinder   4 or 5 minutes   
2  1998-05-15 2

In [14]:
from google.colab import files

# mirem la mida en Megabytes de full.csv i de original.zip
mida_csv = calcular_mida('fullDF.csv')

missatge = (f'La mida del fixter fullDF és {mida_csv}, '
            f'en canvi, la de oiringal.zip és {size_original}')
print(missatge)

# per a fer la comparació en el meu ordinador primer hem de descarregar
# els fixters de la màquina de google

files.download('fullDF.csv')

La mida del fixter fullDF és 113.6, en canvi, la de oiringal.zip és 62.2


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

La diferència de mida entre els formats CSV i ZIP, tot i tindre la mateixa informació es degut al format.

 * format CSV: és un format de fitxer de text pla sense compressió. És a dir, cada caracter dincs del fitxer ocupa un espai en el disc(inclosos els separados i els salts de línia).

 * format ZIP: Aquest suporta la compresió de dades sense pèrdua. Això vol dir que l'algoritme de compresió elimina redundancies del fitxer.

Després de descarregar 'original.zip' i 'fullDF.csv' al meu ordinador (Windows 10) i fer click secundari en ells per veure el tamany en disc he pogut veure que ocupen el mateix que a la màquina de Google.

 * fullDF.csv:  tamaño 114 MB (119.690.349 bytes), tamaño en disco 114 MB (119.693.312 bytes)
 * original.zip: tamaño 62,1 MB (65.200.128 bytes), tamaño disco 62,1 MB (65.200.128 bytes)

 El nombre entre parèntesis està mostrant la mida exacta en bytes, en canvi, els 114 i 62.1 MB estan en notació decimal, on 1 KB = 1000 bytes i 1 MB = 1.000.000 bytes

# Exercici 3

En aquest exercici farem servir el dataframe construït en l'exercici anterior per tal de fer diverses cerques en les nostres dades. A més a més, tornarem a convertir el data frame en una altra estructura de fitxers i veurem si el resultat millora l'us de l'espai de disc que hem observat.

**NOTA** Us recomanem resoldre primer completament l'exercici 2 abans d'intentar l'exercici 3. En tot cas, disposeu d'un petit fitxer anomenat **testDF.csv**. On podeu veure una mostra del tipus de registres que ha de contenir el dataframe que heu de crear a l'exercici 2.

En primer lloc, necessitem explicitar l'any de cada avistament:

# 3.1

Afegiu una columna "year" al dataframe creat en l'exercici 2 que contingui l'any en què es va esdevenir l'avistament.


# 3.2

El cap de l'empresa us diu que sap del cert que només part de les entrades del dataframe són evidència indiscutible de presència extraterrestre: En concret han de complir, que el camp "text" contingui, a la vegada:

- Dues paraules **consecutives**, la primera començant per E i la segona per T (en majúscules).
- Tres paraules **consecutives**, la primera començant per U, la segona per F i la tercera per O (en aquest cas, les lletres inicials poden ser majúscules o minúscules indistintament).

Per exemple, el text

"...Els Tròpics són sempre un festival Olímpic, tot i que hi plou poc..."

Compliria les dues condicions.

Per contra, els textos

"Els vells Tròpics són sempre un festival Olímpic, tot i que hi plou poc"

"Els tròpics són sempre un festival Olímpic, tot i que hi plou poc"

"Els Tròpics són sempre un festival de color, tot i que hi plou poc"

No el compliria donat que: en el primer cas i el segon no hi ha dues paraules consecutives que comencin per E i T. En el primer cas les paraules no són consecutives i en el segon la E i la T no són totes dues majúscules. En el tercer cas, tot i que la primera condició es compleix, la segona no.

**Utilitzant expressions regulars**, creu un nou dataframe (i anomeneu-lo "df32") que contingui només les línies  que compleixin a la vegada les dues condicions esmentades.

# 3.3

Per motius d'espai pot ser necessari reduir la informació que es guarda per a cada  registre. Partint del dataframe resultat de l'exercici 3.1, creeu un nou dataframe anomenat "df33", en el que haureu eliminat la columna "text".

# 3.4

Creeu una funció que rebi com a entrada un string amb la ruta a una carpeta i un dataframe i:


**1) crei un diccionari com el que es detalla a continuació**:

- Les claus del diccionari seran enters representant l'ordre en què cada entrada apareix al dataframe (index).

- El valor associat a cada clau serà, en si mateix, un diccionari amb una clau per a cadascun dels conceptes següents:

summary,city,state,date_time,shape,duration,text,city_latitude,city_longitude, year

i el valor de cada clau segons consta al dataframe l'arxiu que s'està llegint.

    **Pista: Aquest part de l'exercici es pot resoldre amb una sola comanda de la llibreria pandas**

**2) Recorri totes les entrades del diccionari i construeixi una nova estructura de fitxers tal com es detalla a continuació***

- El directori arrel contindrà una subcarpeta per a cada any on s'hagi produït un avistament, a dins de la carpeta de cada any hi haurà carpetes per a tots els estats on hi hagi hagut avistaments i a dins de la carpeta de cada estat tindrem una carpeta per a cada ciutat on hi hagi hagut avistaments. A dins de cada carpeta de ciutat, hi haurà un fitxer de text per a cada tipus d'objecte avistat. Per exemple, "circle.txt", "cylinder.txt" o "cigar.txt".

```
<root_folder>
    <year1>
        <state_code1>
            <city_name1>
                shape1.txt
                shape2.txt
                ...
                shapeN.txt
            <city_name2>
            ...
        <state_code2>
        ...
    <year2>
    ...
            
```

- A cada línia de cada fitxer caldrà escriure-hi la informació referent als següents conceptes (separats per punt i coma ";" ):

summary;date_time;duration;text;city_latitude;city_longitude

- Els noms d'aquests conceptes no apareixeran als fitxers i quedaran implícits per l'ordre a cada línia. Cada fitxer tindrà com a mínim una línia, però pot ser que en tingui més.

- En cas que un concepte no aparegui en un dataframe (per exemple "text" en un dels casos que us demanarem), ignoreu-lo en la llista de conceptes.

**Executeu la funció creada amb els dataframes resultat dels apartats 3.1, 3.2, 3.3.**

# 3.5

- Mesureu i mostreu per pantalla la mida de les tres carpetes resultat de l'apartat 3.3. I guardeu-les en tres variables anomenades "size_31", "size_32", "size_33".
- Compareu les diferències entre aquestes les mides d'aquestes 3 carpetes, i la de la carpeta descomprimida "original" segons les mesura la funció de l'apartat 2.1 i segons les retorna el vostre sistema operatiu i comenteu els resultats en un petit paràgraf de text (<10 linies).

**Nota:** Per saber les mides de fitxers "al sistema operatiu" el més fàcil és que copieu els fitxers i carpetes al vostre drive i els descarregueu per mesurar-los amb les eines del vostre sistema. Aquesta part es considerarà "extra" i es pot obtenir la màxima puntuació a la PAC sense resoldre-la.

<span style="font-family: Courier New; background-color: #ffcc5c; color: #000000; padding: 3px; ">EG</span> **(5 punts)**

**SOLUCIO** En l'apartat 3.1 volem afegir una nova columna al nostre dataframe. La nova columna depèn d'una columna que ja hi ha, així que la transformem a un format que puguem manipular i apliquem la transformació, guardant el resultat en la nova columna.

In [15]:
# fem una copia de df_pac2 anomenada df31
df31 = df_pac2.copy()
# Convertim la columna 'date_time' a datatime
df31['date_time'] = pd.to_datetime(df31['date_time'])
# A continuació obtenim l'any de date_time i l'afegim a la nova columna
df31['year'] = df31['date_time'].dt.year

# mirem les 10 primeres files per veure si s'ha fet correctament
print(df31.head(10))

                                             summary         city state  \
0  Two strobe like lights flying in unison in sou...       Victor    CO   
1                  colored lights behind Pike's Peak       Victor    CO   
2  It was about 11:30 at night, when my brother c...  Pueblo West    CO   
3  two bright white lights with a translucent vei...  Pueblo West    CO   
4  Three flashing lights in a triangle shape over...  Pueblo West    CO   
5  Lights Over Lake Pueblo travel past the North ...  Pueblo West    CO   
6      A red laser-like light came out of a cloud...  Pueblo West    CO   
7  Intense red light, spinning rapidly, droning s...  Pueblo West    CO   
8  Red lights that were hovering in the north.  2...  Pueblo West    CO   
9  I have been watching documentaries and studyin...  Pueblo West    CO   

            date_time     shape         duration  \
0 2012-01-26 19:00:00   unknown       15 minutes   
1 2000-09-24 20:15:00  cylinder   4 or 5 minutes   
2 1998-05-15 23:30

**SOLUCIÓ** En l'arpatat 3.2 volem buscar en el nostre dataframe els registres que a la columna 'text' compleixen unes certes condicions. Primer definim les condicions fent servir expressions regulars. A continuació fem una funció que retorna True si donat un text s'hi troben aquestes condicions. Fem servir la funció per a filtrar el nostre dataframe i crear-ne un de nou.

In [16]:
import re
from typing import List, Tuple

# Expressions regulars
# fem servir r'cadena' per dir que és una cadena 'raw' i poder fer servir
# caràcters com '\' fàcilment
# \b indica principi o final de paraula, al estar davant, indica principi
# \w* indica qualsevol nombre de caràcters alfanumèrics, així tenim en compte
# paraules de longitud 1 o més.
# Més endavant ens preocuparem de les majúscules i les minúscules
re_ET = r'\bE\w* T\w*'
re_ufo = r'\bU\w* F\w* O\w*'


# Funció per buscar expressions regulars en un text
def buscar_condicions(text: str,
                      condicions: List[Tuple[str, bool]]) -> bool:
    """
    Funció per buscar expressions regulars en un text.
    Permet diferenciar entre majúscules i minúscules.

    Args:
    text (str): El text a ser evaluat.
    condicions (List[Tuple[str, bool]]): Llista de tuples amb expressions
    regulars i un booleà indicant si s'ha d'usar IGNORECASE.

    Returns:
    bool: True si es troben totes les condicions, False en cas contrari.
    """
    # itera sobre la llista condicions, cada element és una tupla
    # (condicio, ignore_case) on 'condicio' és una expressió regular
    # i 'ignore_case' és un booleà.
    # Si IGNORECASE està en False, es diferencia entre majúscules i minúscules
    for condicio, ignore_case in condicions:
        if ignore_case:
            # re.search(condicio, text, re.IGNORECASE) busca la condició
            # si no el troba, es fa el return False
            if not re.search(condicio, text, re.IGNORECASE):
                return False
            # fa el mateix però amb ignore_case = false
        else:
            if not re.search(condicio, text):
                return False
    return True


# Com volem que es busqui E i T (majúscules), fem ignore_case <- False
# Com només volem que busqui u, f, o (majúscules i mínuscules) fem True
condicions = [(re_ET, False), (re_ufo, True)]

# Creació del dataframe df32 on apliquem la funció 'buscar_condicions' sobre
# la columna 'text' de 'df_pac2'. Al fer apply es retornen els valors booleans
# assignats a cada fila de la columna text. Aquest resultat es fa servir
# per filtra df31 i posar a df32 només aquelles files que compleixin
# totes les condicions.-
df32 = df31[df31['text'].apply(
    lambda text: buscar_condicions(text, condicions))]

In [17]:
# Test Cell, public
assert len(df32) < 30, "Too many entries"

**SOLUCIÓ** En l'apartat 3.3 volem eliminar la columna 'text' de df31. Per fer-ho hem creat una funció que elimina una de les columnes d'un dataframe i retorna un dataframe, si la columna no existeix, dona un missatge d'error.

In [18]:
def eliminar_column(dataframe: pd.DataFrame, nom_columna: str) -> pd.DataFrame:
    """
    Elimina una columna específica d'un dataframe

    Args:
    dataframe (pd.DataFrame): dataframe a processar
    nom_columna (str): nom de la columna a eliminar

    Returns:
    pd.DataFrame: dataframe amb la columna eliminada o el
                  mateix dataframe amb un missatge d'error
    """
    # Si nom_columna està al df
    if nom_columna in dataframe.columns:
        # eliminem la columna amb .drop
        return dataframe.drop(columns=[nom_columna])
    # en cas contrari
    else:
        # donem el missatge d'error
        print(f"La columna '{nom_columna}' no existeix en el DataFrame.")
        return dataframe


# Intentem eliminar una columna que no existeix
df33 = eliminar_column(df31, 'Teléfono, mi casa')

# Eliminem ara la columna 'text'
df33 = eliminar_column(df31, 'text')

# Mirem si s'ha fet correctament
assert len(df31.columns) == len(df33.columns) + 1, \
       "df31 no té exactament una columna més que df33"

La columna 'Teléfono, mi casa' no existeix en el DataFrame.


**SOLUCIÓ** En l'apartat 3.4 es vol una funció que:

part 1: crei un diccionari amb certes característiques, per fer-ho s'ha fet servir la informació disponible a https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_dict.html.

part 2: Recorri les entrades del diccionari i construeixi una nova estructura de fitxers igual que a la del principi de la pràctica però tenint en compte els anys.


In [19]:
def df_to_diccionari_to_directori(df: pd.DataFrame, directori: str) -> None:
    """
    Processa un dataframe i construeix una estructura de directoris amb
    subcarpetes per any, estat i ciutat. A continuació genera fitxers de
    text per a cada forma d'avistament, finalment hi afegeix la resta
    de la informació.

    Args:
    df (pd.DataFrame): dataFrame amb les dades
    directori (str): ruta de la carpeta on es crearà l'estructura
    """

    # A la llibreria de pandas, en concret en pandas.DataFrame.to_dict he vist
    # que al fer orient='index' s'indica que cada fila del df serà una entrada
    # del diccionari
    diccionari = df.to_dict(orient='index')
    # Recorrem cada element de 'diccionari'
    # idx: és la clau de cada element del diccionari
    # avistament: el valor associat a cada clau
    # es fa servir .items per obtindre les parelles clau-valor
    for idx, avistament in diccionari.items():
        # x = avistament.get('x'): s'extreu el valor associat a 'x',
        # si la clau no existeix retornarà 'None'
        year = avistament.get('year')
        state = avistament.get('state')
        city = avistament.get('city')
        shape = avistament.get('shape')

        # amb os.path.join unim múltiples cadenes de text de forma que
        # és compatible amb el SO utilitzat ('\' Ubuntu, '//' Windows)
        # el directori serà el directori base\any\codi_estat\ciutat
        # year es transforma a string
        dir_path = os.path.join(directori, str(year), state, city)

        # Crea les carpetes de 'dir_path' si no existeixen
        os.makedirs(dir_path, exist_ok=True)

        # Construeix el camí del fitxer i posa el valor de 'shape'
        # com a nom del .txt
        file_path = os.path.join(dir_path, f"{shape}.txt")

        # Creem una llista anomenada info. Cada valor de la llista és un valor
        # obtingut del diccionari. Amb el mètode .get() obtenim la informació
        # i definim un valor per defecte '' en cas de que no hi hagi la clau.
        info = [
            avistament.get('summary', ''),
            str(avistament.get('date_time', '')),
            avistament.get('duration', ''),
            avistament.get('text', ''),
            str(avistament.get('city_latitude', '')),
            str(avistament.get('city_longitude', ''))
        ]
        # unim els element pero els separem amb ';'. Al final fem '\n' per
        # indicar un salt de línia i separar els avistaments
        info_line = ";".join(info) + '\n'

        # Obrim 'file_path' en mode addició, és a dir, si el fixter ja existeix
        # la informació s'afegeix, en cas contrari, crea el fitxer.
        # es fa servir 'with' per asegurar que el fitxer es tanca correctament
        with open(file_path, 'a') as file:
            # escribim la cadena de text 'info_line' al fitxer
            file.write(info_line)

In [20]:
# Ho fem a la màquina local, ja que fer-ho al drive requereix molt de temps
# ruta a la carpeta de la màquina local de Google Collab 'Exercici_3_4_df{xy}'
ruta_local_df31 = 'Exercici_3_4_df31'
ruta_local_df32 = 'Exercici_3_4_df32'
ruta_local_df33 = 'Exercici_3_4_df33'

# fem servir la funció que hem definit sobre els dataframes
df_to_diccionari_to_directori(df31, ruta_local_df31)
df_to_diccionari_to_directori(df32, ruta_local_df32)
df_to_diccionari_to_directori(df33, ruta_local_df33)

In [21]:
# Mirem l'estructura de les carpetes per a buscar possibles errors
mostrar_estructura(ruta_local_df31)
print('\n')
mostrar_estructura(ruta_local_df32)
print('\n')
mostrar_estructura(ruta_local_df33)

Estructura de: Exercici_3_4_df31
2010
  CO
    Pueblo West
      unknown.txt
    Colorado City
      triangle.txt
  MA
    Paxton
      unknown.txt
    Kingston
      triangle.txt
1984
  MA
    Williamstown
      triangle.txt
    Edgartown
      triangle.txt
  WI
    Wabeno
      other.txt
    Columbus
      cigar.txt
Nombre total de fitxers: 88058


Estructura de: Exercici_3_4_df32
1984
  NY
    Ossining
      other.txt
2007
  DC
    Washington
      sphere.txt
  TX
    Dallas
      unknown.txt
Nombre total de fitxers: 15


Estructura de: Exercici_3_4_df33
2010
  CO
    Pueblo West
      unknown.txt
    Colorado City
      triangle.txt
  MA
    Paxton
      unknown.txt
    Kingston
      triangle.txt
1984
  MA
    Williamstown
      triangle.txt
    Edgartown
      triangle.txt
  WI
    Wabeno
      other.txt
    Columbus
      cigar.txt
Nombre total de fitxers: 88058


In [22]:
# Calculem i imprimim els size de les carpetes
size_31 = calcular_mida(ruta_local_df31)
size_32 = calcular_mida(ruta_local_df32)
size_33 = calcular_mida(ruta_local_df33)

print(f'La mida de {ruta_local_df31} és {size_31} Megabytes\n'
      f'La mida de {ruta_local_df32} és {size_32} Megabytes\n'
      f'La mida de {ruta_local_df33} és {size_33} Megabytes\n'
      )

La mida de Exercici_3_4_df31 és 557.1 Megabytes
La mida de Exercici_3_4_df32 és 0.2 Megabytes
La mida de Exercici_3_4_df33 és 69.1 Megabytes



In [23]:
os.listdir()

['.config',
 'df31.zip',
 'Exercici_3_4_df33.zip.zip',
 'original',
 'Exercici_3_4_df33.zip',
 'fullDF.csv',
 'Exercici_3_4_df31.zip.zip',
 'Exercici_3_4_df31.zip',
 'Exercici_3_4_df31',
 'Exercici_3_4_df32.zip',
 'original.zip',
 'Exercici_3_4_df32',
 'df33.zip',
 'Exercici_3_4_df33',
 'df32.zip',
 'drive',
 'Exercici_3_4_df32.zip.zip',
 'sample_data']

In [24]:
# Descarreguem els fitxers previament (comprimits previament)
# i mirem les seves dimensions
import shutil


def comprimir_directori(ruta_directori, nom_fitxer):
    """
    Comprimeix un directori complet en format ZIP
    """
    nom_fitxer_zip = f"{nom_fitxer}.zip"
    shutil.make_archive(nom_fitxer, 'zip', ruta_directori)


# Exemple d'ús de la funció
comprimir_directori(ruta_local_df31, 'df31')
comprimir_directori(ruta_local_df32, 'df32')
comprimir_directori(ruta_local_df33, 'df33')

files.download('df31.zip')
files.download('df32.zip')
files.download('df33.zip')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

**RESUM DE DIMENSIONS**

**Dimensions amb funció de Python:**

original:
- La mida de original és 111.4 Megabbytes

df31:
- La mida de Exercici_3_4_df31 és 334.3 Megabytes

df32:
- La mida de Exercici_3_4_df32 és 0.1 Megabytes

df33:
- La mida de Exercici_3_4_df33 és 27.6 Megabytes

**Dimensions en el meu ordinador (Windows 10):**

original:
- tamaño: 109 MB (114.340.385 bytes)
- tamaño en disco: 187 MB (196.677.632 bytes)

df31:
- tamaño: 409 MB (428.903.716 bytes)
- tamaño en disco: 576 MB (604.565.504 bytes)

df32:
- tamaño: 101 KB (104.196 bytes)
- tamaño en disco: 172 KB (176.128 bytes)

df33:
- tamaño: 72,4 MB (75.968.149 bytes)
- tamaño en disco: 58,9 MB (61.857.792 bytes)

Si comparem les mides dels dataframes entre ells, el més petit és el df32, ja que vam eliminar molts registres al filtrar per les condicions 'ET' i 'ufo'. El següent més petit és el df33, ja que en ell hi tenim tots els registres però hem eliminat la columna 'text'. A continuació tindriem 'original', que tot i tindre menys informació que df31 (ja que a aquest li hem afegit la columna 'year'), el fet d'estar repartit en menys directoris fa que ocupi menys (segurament degut a la dimensió del bloc d'assignació, si el tamany d'un fitxer és menor al de l'assignació d'un bloc, ocupa tot el bloc).

Aquesta forma d'assignar memoria per blocs és el que fa que el tamany i el tamany en disc sigui diferent per directoris pero no per fitxers únics (com el csv o un zip)